### DISCUSSION TITLE
ChanceStyle is a little confusing

### DISCUSSION THREAD

**Discussion by ShuhuaGao** [JuliaReinforcementLearning/ReinforcementLearning.jl#476] [Q&A] [answered]
In the [Guide](https://juliareinforcementlearning.org/docs/How_to_write_a_customized_environment/#[ChanceStyle](@ref)), there is 
>If there's no rng in the environment, everything is deterministic afer taking each action, then we call the ChanceStyle of these environments are of Deterministic. Otherwise, we call them Stochastic

I am confused about the meaning of `ChanceStyle`.
- Does `ChanceStyle` only apply to multi-agent RL? Since the [documentation of ChanceStyle](https://juliareinforcementlearning.org/docs/rlbase/#ReinforcementLearningBase.ChanceStyle-Tuple{T}%20where%20T%3C:AbstractEnv) always mentions a game and a chance player.
- In the above [Guide](https://juliareinforcementlearning.org/docs/How_to_write_a_customized_environment/#[ChanceStyle](@ref)), there is no `rng` field in the `struct LotteryEnv`, but the ChanceStyle shown therein is `Stochastic()`.
- Consider a single-agent RL. If the state transition is deterministic, that is, given `(s, a)`, the resultant `(s', r)` is fixed, what should the ChanceStyle be according to your definition? Such deterministic transition is commonly assumed in practice, like a robot.
-   Does an enviroment's ChanceStyle matter much? Possibly some algorithms will make use of it.

**Comment by findmyway** [comment, marked as answer]
> In the [Guide](https://juliareinforcementlearning.org/docs/How_to_write_a_customized_environment/#%5BChanceStyle%5D(@ref)), there is
> 
> > If there's no rng in the environment, everything is deterministic afer taking each action, then we call the ChanceStyle of these environments are of Deterministic. Otherwise, we call them Stochastic
> 
> I am confused about the meaning of `ChanceStyle`.
> 
> * Does `ChanceStyle` only apply to multi-agent RL? Since the [documentation of ChanceStyle](https://juliareinforcementlearning.org/docs/rlbase/#ReinforcementLearningBase.ChanceStyle-Tuple%7BT%7D%20where%20T%3C:AbstractEnv) always mentions a game and a chance player.

From the API designer's level, no the `ChanceStyle` can be applied to both single and multi agent environments. But in this package, we mainly used it in multi-agent environments.

> * In the above [Guide](https://juliareinforcementlearning.org/docs/How_to_write_a_customized_environment/#%5BChanceStyle%5D(@ref)), there is no `rng` field in the `struct LotteryEnv`, but the ChanceStyle shown therein is `Stochastic()`.

Nice catch! Actually we used the `Random.GLOBAL_RNG` in that game. I didn't add the `rng` field in the definition of that game on purpose to keep that example simple enough for first time users. But yes, to obey the definition of interfaces strickly, we should add a `rng` in that game.

> * Consider a single-agent RL. If the state transition is deterministic, that is, given `(s, a)`, the resultant `(s', r)` is fixed, what should the ChanceStyle be according to your definition? Such deterministic transition is commonly assumed in practice, like a robot.

Then just define it as [DETERMINISTIC](https://juliareinforcementlearning.org/docs/rlbase/#ReinforcementLearningBase.DETERMINISTIC)

> * Does an enviroment's ChanceStyle matter much? Possibly some algorithms will make use of it.

Until now, only some multiaget algorithms require that the games are deterministic or explicit stochastic.

**Reply by ShuhuaGao** [reply]
> to obey the definition of interfaces strickly, we should add a `rng` in that game.

But I do not see where this field `rng` is ever checked for `ChanceStyle`. See the [line](https://github.com/JuliaReinforcementLearning/ReinforcementLearning.jl/blob/196f02a395676bbfb9d2616758e7999e777d4f0e/src/ReinforcementLearningBase/src/interface.jl#L232) below:
```julia
ChanceStyle(::Type{<:AbstractEnv}) = STOCHASTIC
```
Did you miss it in the source code, or is the documentation outdated?

**Reply by findmyway** [reply]
I added some checks here:

https://github.com/JuliaReinforcementLearning/ReinforcementLearning.jl/blob/196f02a395676bbfb9d2616758e7999e777d4f0e/src/ReinforcementLearningBase/src/base.jl#L161-L165

These checks are optional though.

**Reply by ShuhuaGao** [reply]
> But I do not see where this field `rng` is ever checked for `ChanceStyle`

I actually mean with the above words that there is no check like `hasfield(TEnv, :rng)`. Did you refer to the existence of the field `rng` when saying 
>If there's no rng in the environment, everything is deterministic afer taking each action, then we call the ChanceStyle of these environments are of Deterministic. Otherwise, we call them Stochastic

**Reply by findmyway** [reply]
What I mean by the `rng` here is more general than just a `hasfield(Env, :rng)` check. In some cases, the environment is just a wrapper and does not have such a `rng` field. But semantically, environments of `STOCHASTIC` chance style should have implemented the `Random.seed!` interface to allow setting the random seed.

**Reply by ShuhuaGao** [reply]
That makes sense. In that case, the [documentation](https://juliareinforcementlearning.org/docs/How_to_write_a_customized_environment/#[ChanceStyle](@ref)) may better state that a dispatch of `ChanceStyle(::Type{<:AbstractEnv})` should be provided for a custom environment if `ChanceStyle` matters for that environment.

**Reply by findmyway** [reply]
Fair enough, I update that section slightly https://github.com/JuliaReinforcementLearning/ReinforcementLearning.jl/commit/3ccf6edafc9dde4b898555a10198dab3ba517988 . But feel free to make further updates.
